### PR TITLE
Add PPO1

### PR DESCRIPTION
**Pull request by ajaysub110**
Basic implementation is ready. Still to add support for continuous action spaces.

### CODE REVIEW

File: jigglypuffRL/common/policies.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            c = Categorical(probs=x)
+        elif isinstance(self.env.action_space, gym.spaces.Box):
+            mean, log_std = x
+            log_std = torch.clamp(log_std, -20, 2)
+            c = Normal(mean, log_std.exp())
+
+        action = c.sample()
+        return action, c
+
+
```
Comment (Sharad24) [inline jigglypuffRL/common/policies.py]: Can you also define a method for greedy action i.e. just returns the mean of the distribution? For deterministic actions, how are we getting actions?

File: jigglypuffRL/deeprl/ppo1/ppo1.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+        # Assign device
+        if "cuda" in device and torch.cuda.is_available():
+            self.device = torch.device(device)
+        else:
+            self.device = torch.device("cpu")
+
+        # Assign seed
+        self.env.seed(seed)
+        torch.manual_seed(seed)
```
Comment (Sharad24) [inline jigglypuffRL/deeprl/ppo1/ppo1.py]: There are some seed settings missing here right?
```np.random.seed(), random.seed(), torch.backends.cudnn.deterministic = True, torch.backends.cudnn.benchmark = False```
Check https://pytorch.org/docs/stable/notes/randomness.html
Maybe something like this would make more sense:
```if seed!=None:
    torch.manual_seeds()
    np.random....
    .
    .
    .```

File: jigglypuffRL/deeprl/ppo1/ppo1.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        self.policy_old = self.policy_old.to(self.device)
+        self.policy_old.load_state_dict(self.policy_new.state_dict())
+        self.value_fn = get_value_from_name(self.value)(self.env).to(self.device)
+        self.optimizer_policy = opt.Adam(
+            self.policy_new.parameters(), lr=self.lr_policy
+        )
+        self.optimizer_value = opt.Adam(self.value_fn.parameters(), lr=self.lr_value)
+
+    def select_action(self, s):
+        state = torch.from_numpy(s).type(torch.FloatTensor).to(self.device)
```
Comment (Sharad24) [inline jigglypuffRL/deeprl/ppo1/ppo1.py]: To convert to a FloatTensor, you can do ```torch.from_numpy(state).float().to(self.device)``` instead

File: jigglypuffRL/deeprl/ppo1/ppo1.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        gamma=0.99,
+        clip_param=0.2,
+        actor_batch_size=8,
+        epochs=1000,
+        lr_policy=0.001,
+        lr_value=0.005,
+        policy_copy_interval=20,
+        tensorboard_log=None,
+        seed=0,
+        device="cuda",
```
Comment (Sharad24) [inline jigglypuffRL/deeprl/ppo1/ppo1.py]: Should we set default type as ```cpu```? Thinking that when most people clone repo and start using it, they probably wont be testing on GPUs.

File: jigglypuffRL/deeprl/ppo1/ppo1.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        env,
+        timesteps_per_actorbatch=200,
+        gamma=0.99,
+        clip_param=0.2,
+        actor_batch_size=8,
+        epochs=1000,
+        lr_policy=0.001,
+        lr_value=0.005,
+        policy_copy_interval=20,
+        tensorboard_log=None,
```
Comment (Sharad24) [inline jigglypuffRL/deeprl/ppo1/ppo1.py]: Why is default value of ```tensorboard_log=None```? As in, what else can it be other than ```SummaryWriter```?
Maybe it would make more sense to keep it to a boolean value? Depending on whether the user wants tb logging or not

File: jigglypuffRL/deeprl/ppo1/ppo1.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.optim as opt
+from torch.autograd import Variable
+from torch.distributions import Categorical
```
Comment (Sharad24) [inline jigglypuffRL/deeprl/ppo1/ppo1.py]: Will this always be Categorical? It can be other distributions as well right? (not sure)

File: jigglypuffRL/deeprl/ppo1/ppo1.py (hunk 5 above)
Comment (ajaysub110) [reply jigglypuffRL/deeprl/ppo1/ppo1.py]: `tensorboard_log` expects a path (for log dir) as of now. If nothing is provided, logging won't be done. If we set it to a boolean, we will need another parameter for logging location right?

File: jigglypuffRL/common/policies.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    def sample_action(self, x):
+        x = self.forward(x)
+
+        if isinstance(self.env.action_space, gym.spaces.Discrete):
+            x = nn.Softmax(dim=-1)(x)
+            c = Categorical(probs=x)
+        elif isinstance(self.env.action_space, gym.spaces.Box):
+            mean, log_std = x
+            log_std = torch.clamp(log_std, -20, 2)
+            c = Normal(mean, log_std.exp())
```
Comment (Sharad24) [inline jigglypuffRL/common/policies.py]: Maybe it would make more sense to instantiate the distributions in ```___init___``` itself?

File: jigglypuffRL/deeprl/ppo1/ppo1.py (hunk 6 above)
Comment (ajaysub110) [reply jigglypuffRL/deeprl/ppo1/ppo1.py]: Sorry, forgot to remove this from the `ppo1` file. Have moved all the distribution stuff to the policy class

File: jigglypuffRL/deeprl/ppo1/ppo1.py (hunk 5 above)
Comment (Sharad24) [reply jigglypuffRL/deeprl/ppo1/ppo1.py]: Oh ok. I think we will need a logger though, soon. There we can specify the options similar to how stable baselines does it. We can just pass the logger instead of the main class

File: jigglypuffRL/deeprl/ppo1/ppo1.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+                self.policy_old.load_state_dict(self.policy_new.state_dict())
+
+        if self.tensorboard_log:
+            self.writer.close()
+
+
+if __name__ == "__main__":
+    env = gym.make("LunarLander-v2")
+    algo = PPO1("MlpPolicy", "MlpValue", env)
+    algo.learn()
```
Comment (Sharad24) [inline jigglypuffRL/deeprl/ppo1/ppo1.py]: We would probably need a command line parser here, but maybe we can add that later, thats ok for now. For specifying options to train/test, render, etc.

File: jigglypuffRL/deeprl/ppo1/ppo1.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+            if ep % 20 == 0:
+                print("Episode: {}, reward: {}".format(ep, epoch_reward))
+                if self.tensorboard_log:
+                    self.writer.add_scalar("reward", epoch_reward, ep)
+
+            if ep % self.policy_copy_interval == 0:
+                self.policy_old.load_state_dict(self.policy_new.state_dict())
+
+        if self.tensorboard_log:
+            self.writer.close()
```
Comment (Sharad24) [inline jigglypuffRL/deeprl/ppo1/ppo1.py]: `env.render()` missing here?

File: jigglypuffRL/common/policies.py (hunk 1 above)
Comment (ajaysub110) [reply jigglypuffRL/common/policies.py]: Haven't implemented deterministic action yet. Where do you think that should go? Do we add all kinds of action selection to a class called, `GeneralPolicy` and then inherit our policy classes from it (like stable-baselines does)?

File: jigglypuffRL/deeprl/ppo1/ppo1.py (hunk 5 above)
Comment (Sharad24) [reply jigglypuffRL/deeprl/ppo1/ppo1.py]: Will add this to issues

File: jigglypuffRL/common/policies.py (hunk 1 above)
Comment (Sharad24) [reply jigglypuffRL/common/policies.py]: I think I like the current class design you have added a lot.. It keeps the design choice of having Pytorch based models, first. Every env will just call a different distribution which we can do in every class.

Comment (Sharad24) [reply jigglypuffRL/common/policies.py]: Either way, it looks like to me that we can either keep it _Neural Network centric_ or _Env centric_. IMO NN Centric looks much better than Env centric.

File: jigglypuffRL/deeprl/ppo1/ppo1.py (hunk 8 above)
Comment (ajaysub110) [reply jigglypuffRL/deeprl/ppo1/ppo1.py]: Yeah good point. this is just for other's to refer usage while working on other algorithms initially. Will be removed later

File: jigglypuffRL/common/policies.py (hunk 1 above)
Comment (Sharad24) [reply jigglypuffRL/common/policies.py]: So yeah, A `GeneralPolicy` class would make a lot of sense here. I'll add this in issues as well

File: jigglypuffRL/common/policies.py (hunk 7 above)
Comment (ajaysub110) [reply jigglypuffRL/common/policies.py]: The distribution instantiation is depend on outputs from the model. So how exactly do we do that?

**Comment by Sharad24** [comment]
Can you also add wiki on `coding guidelines`, etc on how/what to use for structuring/linting. Code looks really clean which is really good

File: jigglypuffRL/deeprl/ppo1/ppo1.py (hunk 8 above)
Comment (Sharad24) [reply jigglypuffRL/deeprl/ppo1/ppo1.py]: Will resolve this for now. Adding as an issue

File: jigglypuffRL/common/policies.py (hunk 7 above)
Comment (Sharad24) [reply jigglypuffRL/common/policies.py]: Good point. Not sure how to do that

Comment (Sharad24) [reply jigglypuffRL/common/policies.py]: Should we raise this as an issue? We can resolve this for now maybe

**Comment by ajaysub110** [comment]
Sure. I'll do that.
I think we should get back to the deterministic and greedy actions a little later (in another PR) when we are working on the `policies.py` file directly. What do you think?
Apart from that, are we ready to merge this?

File: jigglypuffRL/common/policies.py (hunk 7 above)
Comment (ajaysub110) [reply jigglypuffRL/common/policies.py]: Okay, I'll do that

**Comment by Sharad24** [comment]
I think we can add it to the issue #3
